Enable setting color interpretation

geowurster opened the pull request:
Closes #1133

This PR adds support for setting color interpretation via `ds.set_colorInterp(bids, ColorInterp.enum)` and `$ rio edit-info --colorinterp`.  I also split the `$ rio edit-info` tests into their own module and consolidated some duplicate tests with `@pytest.mark.parametrize()`.

I discovered https://github.com/mapbox/rasterio/issues/100#issuecomment-47825054 as I was opening this PR, and I'm not sure if its still valid.  I seem to be able set color interpretation with the exception of `undefined`, which may be readonly?  I'm seeing some weird behavior that is currently blocked by raising an exception in this PR, but I need to dig into it.

geowurster commented:
This is ready to go.  Looks like GDAL changed how color interpretation information is stored in GeoTIFFs in `2.2.0`: https://trac.osgeo.org/gdal/ticket/6651.  Older versions of GDAL will store this information in the sidecar XML metadata file: https://trac.osgeo.org/gdal/changeset/26674.

sgillies commented:
@geowurster I'd prefer a set-able `colorinterps` property used like

```
>>> dataset.colorinterps
[<ColorInterp.red: 3>, <ColorInterp.green: 4>, <ColorInterp.blue: 5>]
>>> dataset.colorinterps = [ColorInterp.grey for i in src.indexes]
```

I regret the `set_nodatavals()` method and think I should do the same there (different issue).

geowurster commented:
@sgillies Implementing as a `setter` 👍.  Allow me to mull over some implementation options.

I'm on the fence as to how important it is to be able to set color interpretation for a single band, but it does seem a bit more important for `set_nodatavals()`, `set_description()` and `set_units()`, specifically in the context of images storing non-visual information.

sgillies commented:
@geowurster are you still mulling? do you need more feeback from me?

geowurster commented:
@sgillies Ready.

geowurster commented on rasterio/_base.pyx:
```diff
@@ -809,13 +809,78 @@ cdef class DatasetBase(object):
-        band = self.band(bidx)
-        value = GDALGetRasterColorInterpretation(band)
-        return ColorInterp(value)
+        """Returns a sequence of ``ColorInterp.<enum>`` representing
```
Example usage is here.

geowurster commented:
Test failures are caused by GDAL version dependent behavior.  Creating the test file, reopening it, and then setting color interpretation may work.  Whether or not we should guard users against this behavior is a different question.

sgillies commented:
I think the only blocker is this error in the fixture for GDAL versions before 2.2: https://travis-ci.org/mapbox/rasterio/jobs/287988053#L2010

brendan-ward commented on rasterio/_base.pyx:
```diff
@@ -809,13 +809,78 @@ cdef class DatasetBase(object):
+            from collections import OrderedDict
+            with rasterio.open('image.tif') as src:
+                band_mapping = OrderedDict(
+                    zip(src.colorinterp, self.indexes))
```
@geowurster should this be `src.indexes`?

brendan-ward replied:
Also looks like it should be `zip(src.indexes, src.colorinterp))`    (doesn't run as written)

geowurster replied:
@brendan-ward Good catch!

brendan-ward commented on rasterio/_base.pyx:
```diff
@@ -809,13 +809,78 @@ cdef class DatasetBase(object):
+        makes setting color interpretation easier as order is significant.
+
+            from collections import OrderedDict
+            with rasterio.open('image.tif', 'r+') as src:
```
@geowurster I'm not sure about this example - are you envisioning use cases where the user would normally be working with this as a `dict`?

Seems like for example purposes, this could be simplified down to:
```
src.colorinterp = [ColorInterp.blue, ColorInterp.green, ColorInterp.red]
```

geowurster replied:
You're right.  I tend to use as a band mapping, so this example is more opinionated than it needs to be.

brendan-ward commented on rasterio/_base.pyx:
```diff
@@ -809,13 +809,78 @@ cdef class DatasetBase(object):
+                raise RasterioIOError(
+                    "Can only set color interpretation when dataset is "
+                    "opened in 'r+' or 'w' mode, not '{}'.".format(self.mode))
+            elif len(value) != len(self.indexes):
```
`elif` is unnecessary since prior branch always raises an exception.  `if` is sufficient.

brendan-ward commented on rasterio/rio/edit_info.py:
```diff
@@ -94,10 +152,19 @@ def transform_handler(ctx, param, value):
+         "for the same sans alpha band.  This cannot be combined with "
+         "individual band definitions.  Use 'like' to inherit color "
+         "interpretation from '--like'.  Results may be unexpected if the "
+         "input image and template image have a differing number of bands.")
```
Should we raise a warning or an exception in this case?  I'm having a hard time thinking of a case where we'd want to apply color interpretation between two rasters with similar numbers of bands: the only case possibly being going from RGBA to RGB (4 to 3 bands).

brendan-ward commented on tests/test_colorinterp.py:
```diff
@@ -13,13 +19,16 @@ def test_cmyk_interp(tmpdir):
-        assert dst.colorinterp(2) == ColorInterp.magenta
-        assert dst.colorinterp(3) == ColorInterp.yellow
-        assert dst.colorinterp(4) == ColorInterp.black
+        dst.colorinterp = [
```
@geowurster was this supposed to be `==`?  Appears to have no side effects...

brendan-ward commented on tests/test_colorinterp.py:
```diff
@@ -13,13 +19,16 @@ def test_cmyk_interp(tmpdir):
-@pytest.mark.skip(reason="crashing on OS X with Homebrew's GDAL")
+@pytest.mark.skipif(
+    os.environ.get('TRAVIS', os.environ.get('CI', 'false')).lower() != 'true',
+    reason="Crashing on OS X with Homebrew's GDAL")
```
This might need to be revisited.  Worked fine for me with homebrew gdal 2.2.2

geowurster replied:
Turns out this is still a problem, although I altered the `skipif()` so it runs on GDAL `2.2.2`, which also works for me.

brendan-ward commented on tests/test_colorinterp.py:
```diff
@@ -29,6 +38,66 @@ def test_ycbcr_interp(tmpdir):
-        assert dst.colorinterp(1) == ColorInterp.red
-        assert dst.colorinterp(2) == ColorInterp.green
-        assert dst.colorinterp(3) == ColorInterp.blue
+        dst.colorinterp = [
```
Should this be `assert dst.colorinterp ==` ?

geowurster commented:
Addressed feedback.  Still need to fix the GDAL problem.

geowurster commented on .travis.yml:
```diff
@@ -15,6 +15,7 @@ env:
     - GDALVERSION="2.0.3"
     - GDALVERSION="2.1.4"
     - GDALVERSION="2.2.1"
+    - GDALVERSION="2.2.2"
```
Apparently we weren't testing with `2.2.2` yet.

geowurster commented:
@sgillies @brendan-ward Feedback integrated with two small changes: `src.colorinterp` now returns a tuple rather than list, and I added support for `--colorinterp like` and `--all` in `$ rio edit-info`.

I fixed the version dependent GDAL build failure, but the `trunk` build failed with what looks like some slightly different color interpretation handling.  I'm in favor of merging as is and dealing with this condition later, assuming its still present in the next release.  The tests demonstrate that Rasterio is operating properly.

For future debuggers: maybe setting two bands to `gray` is no longer supported?  The first band of the test image is gray and the test iterates through all color interpretations and attempts to set the second band, so at some point the image would have two gray bands and two undefined bands.

https://travis-ci.org/mapbox/rasterio/jobs/293505177

geowurster commented on tests/test_rio_edit_info.py:
```diff
@@ -0,0 +1,509 @@
+        assert src.nodata == 1.0
+
+    templatefile = 'tests/data/RGB.byte.tif'
+    # Dropping '--all' right after 'edit-info' guards against a regression.
```
This was a lurking bug.

geowurster commented on rasterio/_base.pyx:
```diff
@@ -812,13 +812,72 @@ cdef class DatasetBase(object):
+                value = exc_wrap_int(
+                    GDALGetRasterColorInterpretation(self.band(bidx)))
+                out.append(ColorInterp(value))
+            return tuple(out)
```
`src.colorinterp` is now a tuple rather than list.

brendan-ward commented on tests/test_rio_info.py:
```diff
@@ -8,6 +8,361 @@
+    inputfile = str(data.join('RGB.byte.tif'))
+    result = runner.invoke(
+        main_group, ['edit-info', inputfile, '--unset-nodata'])
+    assert result.exit_code == 0
```
In addition to making sure it ran successfully, it would probably be good to check that the output has no metadata value.

geowurster replied:
This only shows up as new due to the test shuffle but I added a check.

brendan-ward commented on tests/test_rio_info.py:
```diff
@@ -8,6 +8,361 @@
+        main_group, ['edit-info', inputfile, '--nodata', 'NaN'])
+    assert result.exit_code == 0
+    with rasterio.open(inputfile) as src:
+        assert src.nodata != src.nodata
```
How does this even work??  As written, this looks like a logic error (how can something not equal itself).  Is there a way to rewrite this to instead make sure it has the desired value set?

geowurster replied:
This only shows up as new due to the test shuffle but its a byproduct of how `numpy.nan` operates.  `np.nan == np.nan == False` and `(np.nan != np.nan) == True`, but I replaced it with a `np.isnan(src.nodata)`, which makes _way_ more sense.

brendan-ward commented on tests/test_rio_info.py:
```diff
@@ -8,6 +8,361 @@
+        'edit-info', inputfile, '--transform', gdal_geotransform])
+    assert result.exit_code != 0
+    assert 'not recognized as an Affine array' in result.output
+    assert gdal_geotransform in result.output
```
Unclear what value this `assert` adds here; the above conditions ensure that it raised an error as expected.

geowurster replied:
This only shows up as new due to the test shuffle but I removed the final `assert`.

geowurster commented on tests/test_rio_edit_info.py:
```diff
@@ -417,3 +420,41 @@ def test_colorinterp_bad_name():
+    the target image raises an exception.
+    """
+    # Isolate to avoid potential '.aux.xml' creation.
+    rgba = str(data.join('RGBA.byte.tif'))
```
Those fixtures provide explicit paths to `tests/data/RGB[A].tif` and the `data` fixture provides totally new and isolated copies of these files.  An unnoticed or unexpected PAM file (`.aux.xml`) can make debugging or implementing a new feature painful so using isolated files is safer.  In this particular test it shouldn't happen, but a change to `$ rio edit-info` could trigger a PAM file to be created and alter the behavior of tests that point to this file.

The real solution is to make all of our fixtures copy the test files every time they execute, and I opened https://github.com/mapbox/rasterio/issues/1187 to handle that.

geowurster commented:
@brendan-ward Thanks for looking this over again.  I wrote most of this one in the middle of the night, so extra 👀 is 👍.  As you mentioned, a few of your comments are unrelated to this PR, but they were easy to fix.

sgillies commented on docs/cli.rst:
```diff
@@ -259,6 +260,21 @@ or set its nodata value to, e.g., `0`:
+
+.. code-block:: console
+
+    $ rio edit-info --colorinterp 1=red,2=green,3=blue,4=alpha example.tif
```
@geowurster my last question: would you be willing to consider a syntax change here?

rio-edit-info already allows us to set units and descriptions of selected bands like

`rio edit-info --bidx 1 --description "Band 1"`

I'd prefer if color interpretation were done more like that. For example:

`rio edit-info --bidx 1 --colorinterp red`

and for multiple bands in one go

`rio edit-info --bidx 1,2,3 --colorinterp red,green,blue`

I think we absolutely need to keep the shortcuts. That would be my go-to usage.

geowurster replied:
@sgillies That makes commands like `$ rio edit-info --bidx 1,2,3 --description whatever --colorinterp red,green,blue` difficult.  What about dropping the `bidx=ci` syntax and just using `--colorinterp red,green,blue` and requiring the user to specify color interpretations for every band like `ds.colorinterp = [ColorInterp.red, ColorInterp.green, ColorInterp.blue]`?  The `bidx=ci` syntax is a hold over from when I was initially thinking about this in terms of a band mapping rather than a sequence.

sgillies replied:
@geowurster I like dropping the `bidx=ci` syntax. How about the option to "broadcast" a single value across all bands like `--colorinterp grey` ? I'm happy to do that in a different PR.

geowurster replied:
@sgillies Switched syntaxes.  Lets handle broadcasting in a different PR.
